ppx/browser/ppx_deriving_json_runtime.ml

```diff
@@ -6,6 +6,100 @@ let to_string t = Js.Json.stringify t
 
 exception Of_string_error of string
 
+module Classify = struct
+  let classify :
+      t ->
+      [ `Null
+      | `String of string
+      | `Float of float
+      | `Int of int
+      | `Bool of bool
+      | `List of t list
+      | `Assoc of (string * t) list ] =
+   fun json ->
+    if (Obj.magic json : 'a Js.null) == Js.null then `Null
+    else
+      match Js.typeof json with
+      | "string" -> `String (Obj.magic json : string)
+      | "number" ->
+          let v = (Obj.magic json : float) in
+          if Js.Float.isFinite v && Js.Math.floor_float v == v then
+            `Int (Obj.magic v : int)
+          else `Float v
+      | "boolean" -> `Bool (Obj.magic json : bool)
+      | "object" ->
+          if Js.Array.isArray json then
+            let xs = Array.to_list (Obj.magic json : t array) in
+            `List xs
+          else
+            let xs = Js.Dict.entries (Obj.magic json : t Js.Dict.t) in
+            `Assoc (Array.to_list xs)
+      | typ -> failwith ("unknown JSON value type: " ^ typ)
+end
+
+let show_json_error ?depth json =
+  let buffer = Buffer.create 1 in
+  let rec loop ?depth json =
+    let json = Classify.classify json in
+    let depth = Option.map (fun i -> i - 1) depth in
+    match depth with
+    | Some 0 -> Buffer.add_string buffer "_"
+    | _ -> (
+        match json with
+        | `Assoc assoc ->
+            Buffer.add_string buffer "{";
+            List.iter
+              (fun (k, v) ->
+                Buffer.add_string buffer {|"|};
+                Buffer.add_string buffer k;
+                Buffer.add_string buffer {|": |};
+                loop ?depth v;
+                Buffer.add_string buffer {|, |})
+              assoc;
+            Buffer.add_string buffer "}"
+        | `Bool bool ->
+            Buffer.add_string buffer (if bool then "true" else "false")
+        | `Float float -> Buffer.add_string buffer (string_of_float float)
+        | `Int int -> Buffer.add_string buffer (string_of_int int)
+        | `List li ->
+            Buffer.add_string buffer "[";
+            List.iter
+              (fun elt ->
+                loop ?depth elt;
+                Buffer.add_string buffer ", ")
+              li;
+            Buffer.add_string buffer "]"
+        | `Null -> Buffer.add_string buffer "null"
+        | `String str ->
+            Buffer.add_string buffer {|"|};
+            Buffer.add_string buffer (String.escaped str);
+            Buffer.add_string buffer {|"|})
+  in
+
+  (loop ?depth:(Option.map (fun i -> i + 1) depth)) json;
+  Buffer.contents buffer
+
+exception Of_json_error = Json.Decode.DecodeError
+
+let of_json_msg_error msg = raise (Of_json_error (Json_error msg))
+
+let of_json_error ?(depth = 2) ~json msg =
+  of_json_msg_error
+    (Printf.sprintf "%s but got %s" msg (show_json_error ~depth json))
```

Comment:
we need to get rid of `Printf.sprintf` in browser runtime (I know it's used now but I got rid of it in #45)